Source/Core/Common/SmallVector.h

```diff
@@ -14,39 +14,82 @@ namespace Common
 template <typename T, size_t MaxSize>
 class SmallVector final
 {
-  static_assert(std::is_standard_layout_v<T> == true, "Type must be a standard layout type");
-
 public:
+  using value_type = T;
+
   SmallVector() = default;
-  explicit SmallVector(size_t size) : m_size(size) {}
+  explicit SmallVector(size_t new_size) { resize(new_size); }
+
+  ~SmallVector() { clear(); }
+
+  SmallVector(const SmallVector& other)
+  {
+    for (auto& value : other)
+      emplace_back(value);
+  }
 
-  void push_back(const T& x) { m_array[m_size++] = x; }
-  void push_back(T&& x) { m_array[m_size++] = std::move(x); }
+  SmallVector& operator=(const SmallVector& rhs)
+  {
+    clear();
+    for (auto& value : rhs)
+      emplace_back(value);
+  }
+
+  SmallVector(SmallVector&& other)
+  {
+    for (auto& value : other)
+      emplace_back(std::move(value));
+    other.clear();
+  }
+
+  SmallVector& operator=(SmallVector&& rhs)
+  {
+    clear();
```

Comment:
Could we do something like:

```
swap(rhs);
rhs.clear();
```

And implement swap like:

```
void swap(SmallVector& other)
{
  std::swap(m_data, other.m_data);
  m_size = other.m_size;
}
```

Not sure if this would be a tiny bit faster?

Reply:
`swap`ing the underlying buffer like that won't always be safe. `m_array` is now basically a char array, swapping won't call the appropriate copy/move operations of `value_type`.

Reply:
Yeah, looking a bit more, I see now it is undefined behavior 😮 .  Thanks.